File: src/runtime/config/schemas/triattention.cpp
// Registration for the "triattention" namespace schema.
//
// Mirrors python/tensorrt_model_connect/runtime_config/schemas/triattention.py.
// Only edit both sides together; the cross-language match test gates on it.

#include "trtmc/config/schemas/triattention.h"

#include <any>
#include <cstdint>
#include <set>
#include <string>

namespace trtmc::config::schemas {

namespace {

const std::set<Layer> kBundleAndSession = {
    Layer::BuildTime,
    Layer::BundleDefault,
    Layer::PlatformProfile,
    Layer::SessionRequest,
};
const std::set<Layer> kSessionOnly = {Layer::SessionRequest};
const std::set<Layer> kBuildOnly = {Layer::BuildTime, Layer::BundleDefault};

bool is_positive_int32(const std::any& v) {
    if (v.type() != typeid(std::int32_t))
        return false;
    return std::any_cast<std::int32_t>(v) > 0;
}

bool is_nonnegative_int32(const std::any& v) {
    if (v.type() != typeid(std::int32_t))
        return false;
    return std::any_cast<std::int32_t>(v) >= 0;
}

bool is_ge1_int32(const std::any& v) {
    if (v.type() != typeid(std::int32_t))
        return false;
    return std::any_cast<std::int32_t>(v) >= 1;
}

bool is_mean_or_max(const std::any& v) {
    if (v.type() != typeid(std::string))
        return false;
    const auto& s = std::any_cast<const std::string&>(v);
    return s == "mean" || s == "max";
}

ConfigField bool_field(const std::string& name, bool default_value, const std::set<Layer>& layers) {
    return ConfigField{name, "bool", std::any{default_value}, layers, nullptr};
}

ConfigField int_field(const std::string& name, std::int32_t default_value,
                      const std::set<Layer>& layers,
                      std::function<bool(const std::any&)> validator = nullptr) {
    return ConfigField{name, "int32", std::any{default_value}, layers, std::move(validator)};
}

ConfigField str_field(const std::string& name, const std::string& default_value,
                      const std::set<Layer>& layers,
                      std::function<bool(const std::any&)> validator = nullptr) {
    return ConfigField{name, "string", std::any{default_value}, layers, std::move(validator)};
}

} // namespace

Schema make_triattention_schema() {
    return Schema{
        "triattention",
        {
            // --- Core runtime config (formerly bundle + overrides) ------
            bool_field("enabled", false, kBundleAndSession),
            int_field("kv_budget", 4096, kBundleAndSession, is_positive_int32),
            int_field("divide_length", 128, kBundleAndSession, is_positive_int32),
            int_field("recent_window", 128, kBundleAndSession, is_nonnegative_int32),
            str_field("score_aggregation", "mean", kBundleAndSession, is_mean_or_max),
            str_field("per_layer_aggregation", "mean", kBundleAndSession, is_mean_or_max),
            bool_field("count_prompt_tokens", true, kBundleAndSession),
            bool_field("protect_prefill", true, kBundleAndSession),
            bool_field("disable_mlr", false, kBundleAndSession),
            bool_field("disable_trig", false, kBundleAndSession),
            int_field("offset_max_length", 65536, kBundleAndSession, is_positive_int32),
            str_field("stats_section", "triattention_stats.json", kBuildOnly),
            // --- Debug / profiling (session-only) -----------------------
            bool_field("debug", false, kSessionOnly),
            bool_field("profile", false, kSessionOnly),
            int_field("runtime_bucket_rows", 32, kSessionOnly, is_ge1_int32),
            bool_field("disable_gpu_selection", false, kSessionOnly),
            bool_field("disable_gpu_compaction", false, kSessionOnly),
            bool_field("disable_gpu_state", false, kSessionOnly),
            bool_field("zero_tail", false, kSessionOnly),
            str_field("dump_keep_path", "", kSessionOnly),
            int_field("dump_compaction_index", 0, kSessionOnly, is_nonnegative_int32),
            bool_field("abort_after_dump", false, kSessionOnly),
            bool_field("dump_score_cache", false, kSessionOnly),
            bool_field("dump_score_values", false, kSessionOnly),
        },
    };
}

} // namespace trtmc::config::schemas

namespace trtmc::config::schemas {
REGISTER_CONFIG_SCHEMA_FACTORY_WITH_MANIFEST(register_triattention_schema,
                                             make_triattention_schema);
} // namespace trtmc::config::schemas

File: python/tensorrt_model_connect/runtime_config/schemas/triattention.py
"""Schema for the ``triattention`` namespace.

Replaces the former ``TRTMC_TRIATTN_*`` env vars and the fragment of bundle
``config.json`` previously parsed by
family-local ``src/runtime/models/<family>/triattention_kv_cache.cpp``
parsers.

Each field declares which config layers may supply a value. The former
``TRTMC_TRIATTN_OVERRIDE_*`` env vars (they baked the "override" frame into
their names) become layer-gated fields here — the same value shape, but
now ``SESSION_REQUEST`` is just a legitimate source, not an "override".
"""

from __future__ import annotations

from tensorrt_model_connect.runtime_config import (
    ConfigField,
    Layer,
    Schema,
    register_schema,
)


_BUNDLE_AND_SESSION = frozenset({
    Layer.BUILD_TIME,
    Layer.BUNDLE_DEFAULT,
    Layer.PLATFORM_PROFILE,
    Layer.SESSION_REQUEST,
})
_SESSION_ONLY = frozenset({Layer.SESSION_REQUEST})


def _agg_validator(v) -> bool:
    return isinstance(v, str) and v in {"mean", "max"}


SCHEMA = Schema(
    namespace="triattention",
    fields=(
        # --- Core runtime config (formerly bundle config.json + overrides) ---
        ConfigField(
            name="enabled",
            type_tag="bool",
            default=False,
            allowed_layers=_BUNDLE_AND_SESSION,
        ),
        ConfigField(
            name="kv_budget",
            type_tag="int32",
            default=4096,
            allowed_layers=_BUNDLE_AND_SESSION,
            validator=lambda v: isinstance(v, int) and v > 0,
        ),
        ConfigField(
            name="divide_length",
            type_tag="int32",
            default=128,
            allowed_layers=_BUNDLE_AND_SESSION,
            validator=lambda v: isinstance(v, int) and v > 0,
        ),
        ConfigField(
            name="recent_window",
            type_tag="int32",
            default=128,
            allowed_layers=_BUNDLE_AND_SESSION,
            validator=lambda v: isinstance(v, int) and v >= 0,
        ),
        ConfigField(
            name="score_aggregation",
            type_tag="string",
            default="mean",
            allowed_layers=_BUNDLE_AND_SESSION,
            validator=_agg_validator,
        ),
        ConfigField(
            name="per_layer_aggregation",
            type_tag="string",
            default="mean",
            allowed_layers=_BUNDLE_AND_SESSION,
            validator=_agg_validator,
        ),
        ConfigField(
            name="count_prompt_tokens",
            type_tag="bool",
            default=True,
            allowed_layers=_BUNDLE_AND_SESSION,
        ),
        ConfigField(
            name="protect_prefill",
            type_tag="bool",
            default=True,
            allowed_layers=_BUNDLE_AND_SESSION,
        ),
        ConfigField(
            name="disable_mlr",
            type_tag="bool",
            default=False,
            allowed_layers=_BUNDLE_AND_SESSION,
        ),
        ConfigField(
            name="disable_trig",
            type_tag="bool",
            default=False,
            allowed_layers=_BUNDLE_AND_SESSION,
        ),
        ConfigField(
            name="offset_max_length",
            type_tag="int32",
            default=65536,
            allowed_layers=_BUNDLE_AND_SESSION,
            validator=lambda v: isinstance(v, int) and v > 0,
        ),
        # Build-time only: points to a bundle section with calibration stats.
        ConfigField(
            name="stats_section",
            type_tag="string",
            default="triattention_stats.json",
            allowed_layers=frozenset({Layer.BUILD_TIME, Layer.BUNDLE_DEFAULT}),
        ),
        # --- Debug / profiling knobs (session-only) ---
        ConfigField(
            name="debug",
            type_tag="bool",
            default=False,
            allowed_layers=_SESSION_ONLY,
        ),
        ConfigField(
            name="profile",
            type_tag="bool",
            default=False,
            allowed_layers=_SESSION_ONLY,
        ),
        ConfigField(
            name="runtime_bucket_rows",
            type_tag="int32",
            default=32,
            allowed_layers=_SESSION_ONLY,
            validator=lambda v: isinstance(v, int) and v >= 1,
        ),
        ConfigField(
            name="disable_gpu_selection",
            type_tag="bool",
            default=False,
            allowed_layers=_SESSION_ONLY,
        ),
        ConfigField(
            name="disable_gpu_compaction",
            type_tag="bool",
            default=False,
            allowed_layers=_SESSION_ONLY,
        ),
        ConfigField(
            name="disable_gpu_state",
            type_tag="bool",
            default=False,
            allowed_layers=_SESSION_ONLY,
        ),
        ConfigField(
            name="zero_tail",
            type_tag="bool",
            default=False,
            allowed_layers=_SESSION_ONLY,
        ),
        ConfigField(
            name="dump_keep_path",
            type_tag="string",
            default="",
            allowed_layers=_SESSION_ONLY,
        ),
        ConfigField(
            name="dump_compaction_index",
            type_tag="int32",
            default=0,
            allowed_layers=_SESSION_ONLY,
            validator=lambda v: isinstance(v, int) and v >= 0,
        ),
        ConfigField(
            name="abort_after_dump",
            type_tag="bool",
            default=False,
            allowed_layers=_SESSION_ONLY,
        ),
        ConfigField(
            name="dump_score_cache",
            type_tag="bool",
            default=False,
            allowed_layers=_SESSION_ONLY,
        ),
        ConfigField(
            name="dump_score_values",
            type_tag="bool",
            default=False,
            allowed_layers=_SESSION_ONLY,
        ),
    ),
)


register_schema(SCHEMA)
